server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -150,9 +169,52 @@ private void persistRepositories(DbSession dbSession, List<RulesDefinition.Repos
     dbSession.commit();
   }
 
-  @Override
-  public void stop() {
-    // nothing
+
+  private void synchronizeDeprecatedKeys(DbSession dbSession, List<RulesDefinition.ExtendedRepository> repositories) {
+    Set<SingleDeprecatedRuleKey> currentDeprecatedKeys = dbClient.ruleDao().selectAllDeprecatedRuleKeys(dbSession)
+      .stream()
+      .map(SingleDeprecatedRuleKey::from)
+      .collect(MoreCollectors.toSet());
+
+    Set<SingleDeprecatedRuleKey> newDeprecatedKeys = repositories.stream()
+      .flatMap(r -> r.rules().stream())
+      .flatMap(rule -> SingleDeprecatedRuleKey.from(rule).stream())
+      .collect(MoreCollectors.toSet());
+
+    // DeprecatedKeys that must be deleted
+    List<String> uuidsToBeDeleted = difference(currentDeprecatedKeys, newDeprecatedKeys).stream()
+      .map(SingleDeprecatedRuleKey::getUuid)
+      .collect(MoreCollectors.toList());
+
+    dbClient.ruleDao().deleteDeprecatedRuleKeys(dbSession, uuidsToBeDeleted);
+
+    // DeprecatedKeys that must be created
+    Sets.SetView<SingleDeprecatedRuleKey> deprecatedRuleKeysToBeCreated = difference(newDeprecatedKeys, currentDeprecatedKeys);
+
+    Set<RuleKey> ruleKeys = deprecatedRuleKeysToBeCreated.stream()
+      .map(d -> RuleKey.of(d.getOldRepositoryKey(), d.getOldRuleKey()))
+      .collect(MoreCollectors.toSet(deprecatedRuleKeysToBeCreated.size()));
+
+    Map<RuleKey, Integer> ruleKeyToIdMap = dbClient.ruleDao().selectDefinitionByKeys(dbSession, ruleKeys).stream()
+      .collect(MoreCollectors.uniqueIndex(RuleDefinitionDto::getKey, RuleDefinitionDto::getId));
+
+    deprecatedRuleKeysToBeCreated
+      .forEach(r -> {
+        Integer oldId = ruleKeyToIdMap.get(r.getOldRuleKeyAsRuleKey());
+        if (oldId == null) {
+          return;
+        }
+
+        dbClient.ruleDao().insert(dbSession, new DeprecatedRuleKeyDto()
+          .setUuid(uuidFactory.create())
+          .setRuleId(oldId)
+          .setOldRepositoryKey(r.getOldRepositoryKey())
+          .setOldRuleKey(r.getOldRuleKey())
+          .setCreatedAt(system2.now())
+        );
+
+        // FIXME Here the renaming of the KEY ?
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Take the required action to fix the issue indicated by this comment. ([squid:S1134](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10311&issues=AWFRf5ZjsS4rYBZqFGC5&open=AWFRf5ZjsS4rYBZqFGC5)

[ISSUE_KEY]: AWFRf5ZjsS4rYBZqFGC5